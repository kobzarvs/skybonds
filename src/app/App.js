import { BondCard } from '../features/bond-card/ui';
import { createDataset } from '../features/bond-card/db/bonds-factory';
import { useEffect, useState } from 'react';
import {
    Affix,
    AppShell,
    Burger, Button,
    Group,
    NavLink, rem,
    Title, Transition,
} from '@mantine/core';
import { useDisclosure, useWindowScroll } from '@mantine/hooks';
import { IconArrowUp } from '@tabler/icons-react';
import { BondCachePage } from '../features/bonds-cache/ui';


const pages = [
    {
        name: 'Bond Card',
        description: 'Карточка облигации',
    },
    {
        name: 'Bond Cache',
        description: 'Кэш данных об облигациях',
    },
    {
        name: 'Shared Construction',
        description: 'Долевое строительство',
    },
];

function App() {
    const [pending, setPending] = useState(false);
    const [updatedAt, update] = useState(Date.now());

    let _pageName;
    try {
        _pageName = localStorage.getItem('bonds-page-name');
    } catch (e) {
        console.error(e);
    }
    const [pageName, setPageName] = useState(_pageName ?? pages[0].name);

    const [scroll, scrollTo] = useWindowScroll();
    const [mobileOpened, { toggle: toggleMobile }] = useDisclosure();
    const [desktopOpened, { toggle: toggleDesktop }] = useDisclosure(true);

    useEffect(() => {
        try {
            localStorage.setItem('bonds-page-name', pageName);
        } catch (e) {
            console.error(e);
        }
    }, [pageName]);

    let page;
    switch (pageName) {
        case 'Bond Card':
            page = <BondCard updatedAt={updatedAt}/>;
            break;
        case 'Bond Cache':
            page = <BondCachePage/>;
            break;
        case 'Shared Construction':
            break;
        default:
            page = <BondCard updatedAt={updatedAt}/>;
    }

    return (
        <AppShell
            header={{ height: 60 }}
            navbar={{
                width: 300,
                breakpoint: 'sm',
                collapsed: {
                    mobile: !mobileOpened,
                    desktop: !desktopOpened,
                },
            }}
            padding="md"
        >
            <AppShell.Header>
                <Group h="100%" px="md" justify="space-between" align="center">
                    <Group>
                        <Burger opened={mobileOpened}
                                onClick={toggleMobile}
                                hiddenFrom="sm"
                                size="sm"
                        />
                        <Burger opened={desktopOpened}
                                onClick={toggleDesktop}
                                visibleFrom="sm"
                                size="sm"
                        />
                        <Title>{pageName}</Title>
                    </Group>
                    <Button
                        loading={pending}
                        onClick={async () => {
                            if (pending) return;
                            setPending(true);
                            setTimeout(async () => {
                                const time = await createDataset(true);
                                update(time);
                                setPending(false);
                            }, 150);
                        }}
                    >
                        New dataset
                    </Button>
                </Group>
            </AppShell.Header>
            <AppShell.Navbar p="md">
                Pages
                {pages.map((item, index) => (
                    <NavLink key={item.name}
                             label={item.name}
                             description={item.description}
                             active={pageName === item.name}
                             onClick={() => setPageName(item.name)}
                    />
                ))}
            </AppShell.Navbar>
            <AppShell.Main>
                {page}
            </AppShell.Main>

            <Affix position={{
                bottom: 20,
                right: 20,
            }}
            >
                <Transition transition="slide-up" mounted={scroll.y > 0}>
                    {(transitionStyles) => (
                        <Button
                            leftSection={<IconArrowUp style={{
                                width: rem(16),
                                height: rem(16),
                            }}
                            />}
                            style={transitionStyles}
                            onClick={() => scrollTo({ y: 0 })}
                        >
                            Scroll to top
                        </Button>
                    )}
                </Transition>
            </Affix>
        </AppShell>
    );
}

export default App;
