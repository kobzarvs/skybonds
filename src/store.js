import { configureStore } from '@reduxjs/toolkit';
import { bondsApi } from './features/bond-card/services/bondsApi';
import { setupListeners } from '@reduxjs/toolkit/query';
import { bondsSlice } from './features/bond-card/services/bondsSlice';

export const store = configureStore({
    reducer: {
        bonds: bondsSlice.reducer,
        [bondsApi.reducerPath]: bondsApi.reducer,
    },
    middleware: (getDefaultMiddleware) => {
        return getDefaultMiddleware().concat(bondsApi.middleware);
    },
});

setupListeners(store.dispatch);
