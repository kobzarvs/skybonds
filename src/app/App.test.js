import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import App from './App';
import { store } from '../store';


jest.mock('echarts-for-react', () => {
    return {
        __esModule: true,
        default: () => {
            return <div>Mocked ReactECharts</div>;
        },
    };
});

test('renders learn react link', () => {
    const result = renderer.create(
        <Provider store={store}>
            <App/>
        </Provider>,
    );
    const app = result.toJSON();
    expect(app).toMatchSnapshot();
});
