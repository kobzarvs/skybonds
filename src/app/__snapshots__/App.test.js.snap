// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link 1`] = `
<div
  className="App"
>
  <header
    className="header"
  >
    <span>
      Bond Card
    </span>
    <button
      className="resetButton"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Generate dataset
    </button>
  </header>
  <main
    className="main"
  >
    <div
      className="card"
    >
      <header>
        <h1
          className="header"
        >
          <span>
            Bond A
          </span>
          <span>
            <span
              data-dir="up"
            >
              $5,355.90
            </span>
          </span>
        </h1>
      </header>
      <div
        className="controls"
      >
        <label>
          Select Bond:  
          <select
            onChange={[Function]}
            value="RU000A0JX0J2"
          >
            <option
              value="RU000A0JX0J2"
            >
              Bond A
            </option>
            <option
              value="XS0971721963"
            >
              Bond B
            </option>
            <option
              value="RU000A0JU4L3"
            >
              Bond C
            </option>
            <option
              value="XS0971712345"
            >
              Bond D
            </option>
            <option
              value="XS0971798765"
            >
              Bond E
            </option>
          </select>
        </label>
        <div>
          <button
            className="button"
            data-active={false}
            onClick={[Function]}
          >
            Week
          </button>
          <button
            className="button"
            data-active={true}
            onClick={[Function]}
          >
            Month
          </button>
          <button
            className="button"
            data-active={false}
            onClick={[Function]}
          >
            Quarter
          </button>
          <button
            className="button"
            data-active={false}
            onClick={[Function]}
          >
            Year
          </button>
          <button
            className="button"
            data-active={false}
            onClick={[Function]}
          >
            Maximum
          </button>
        </div>
      </div>
      <div>
        Mocked ReactECharts
      </div>
      <select
        className="parameters"
        onChange={[Function]}
        value="price"
      >
        <option
          value="price"
        >
          Price
        </option>
        <option
          value="yield"
        >
          Yield
        </option>
        <option
          value="spread"
        >
          Spread
        </option>
        <option
          value="volume"
        >
          Volume
        </option>
      </select>
      <div
        className="paginator"
      >
        <button
          onClick={[Function]}
          style={
            Object {
              "width": 40,
            }
          }
          type="button"
        >
          &lt;
        </button>
        <label>
          Page  
          <select
            onChange={[Function]}
            style={
              Object {
                "minWidth": 60,
              }
            }
            value={0}
          />
        </label>
        <span>
          of   
          0
        </span>
        <button
          onClick={[Function]}
          style={
            Object {
              "width": 40,
            }
          }
          type="button"
        >
          &gt;
        </button>
        <label>
          Page size  
          <select
            onChange={[Function]}
            value={10}
          >
            <option
              value="10"
            >
              10
            </option>
            <option
              value="25"
            >
              25
            </option>
            <option
              value="50"
            >
              50
            </option>
          </select>
        </label>
      </div>
      <h1>
        No data
      </h1>
    </div>
  </main>
</div>
`;
